### PR TITLE
SnapStart integration enhancement

### PR DESCRIPTION
**Pull request by cescoffier**
- Write documentation
- Use SnapStart instead of CRaC to avoid the confusion

### CODE REVIEW

File: docs/src/main/asciidoc/amazon-snapstart.adoc [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+You can disable this behavior using:
+
+[source, properties]
+----
+quarkus.snap-start.initialize-classes=false
+----
+
+== Client Priming
+
+Client _priming_ is a technic that allows initializing a client during the snapshotting process, so, it's already fully functional during the application runtime.
```
Comment (maxandersen) [inline docs/src/main/asciidoc/amazon-snapstart.adoc:109]:
```suggestion
Client _priming_ is a technique that allows initializing a client during the snapshotting process, so, it's already fully functional during the application runtime.
```

File: docs/src/main/asciidoc/amazon-snapstart.adoc [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+
+There are two ways to achieving _priming_:
```
Comment (maxandersen) [inline docs/src/main/asciidoc/amazon-snapstart.adoc:111]:
```suggestion
There are two ways to achieve _priming_:
```

File: docs/src/main/asciidoc/amazon-snapstart.adoc [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+
+1. initialize the client in a `static` block, which, thanks to class preloading will be executed before the snapshot
```
Comment (maxandersen) [inline docs/src/main/asciidoc/amazon-snapstart.adoc:113]: should we warn or be explicit that if one does too much here your app might not build with native?

File: docs/src/main/asciidoc/amazon-snapstart.adoc [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+    }
+    // ...
+}
+----
+
+The next section covers (2).
+
+== Resource registration
+
+SnapStart uses CRaC API to allow the application to execute custom code before the snapshotting or during the restoration.
```
Comment (gsmet) [inline docs/src/main/asciidoc/amazon-snapstart.adoc:139]:
```suggestion
SnapStart uses the CRaC API to allow the application to execute custom code before the snapshotting or during the restoration.
```

File: docs/src/main/asciidoc/amazon-snapstart.adoc [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+
+IMPORTANT: While it's the CRaC API, SnapStart is not CRaC and can do _things_ that would not work with others CRaC implementations.
+
+[source, java]
+----
+package me.escoffier.hello;
```
Comment (gsmet) [inline docs/src/main/asciidoc/amazon-snapstart.adoc:145]: I think we should stay away of using your name in examples? I think we use `org.acme` elsewhere.

File: docs/src/main/asciidoc/amazon-snapstart.adoc [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+
+    @Override
+    public void afterRestore(Context<? extends Resource> context) throws Exception {
+        logger.info("after checkout hook");
+        // if there is anything to do during the restoration, do it here.
+    }
+}
+----
+
+WARNING: Restoration is limited to 2s.
```
Comment (gsmet) [inline docs/src/main/asciidoc/amazon-snapstart.adoc:184]:
```suggestion
WARNING: Restoration is limited to 2 seconds.
```

**Review by gsmet — COMMENTED** [review]
I added a few suggestions and comments.

File: docs/src/main/asciidoc/amazon-snapstart.adoc (hunk 3 above)
Comment (cescoffier) [reply docs/src/main/asciidoc/amazon-snapstart.adoc:113]: Oh yeah! I forgot to mention that!

File: docs/src/main/asciidoc/amazon-snapstart.adoc [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+Please check the AWS documentation to verify the eligibility of your AWS region.
+
+== Enabling / Disabling SnapStart Optimizations
+
+If you use the Quarkus AWS Lambda extension, SnapStart optimizations are automatically enabled.
+However, you can enable/disable it explicitly using:
+
+[source, properties]
+----
+quarkus.snap-start.enabled=true|false
```
[the remark this answers is not in the document]
Comment (cescoffier) [reply docs/src/main/asciidoc/amazon-snapstart.adoc:25]: Yeah, I also think `snapstart` is better.

**Review by geoand — APPROVED** [review]

**Review by gsmet — APPROVED** [review]

**Comment by geoand** [comment]
Very good idea to break Crac and Snapstart up in the code.

**Comment by loicmathieu** [comment]
SnapStart is only for Amazon Lambda right ?
Maybe it should be moved to the Amazon Lambda extension.

My understanding is that CRaC support was removed now, it's a little shame as it was promising but I understand that such early tech is not yet supported.

**Comment by cescoffier** [comment]
@loicmathieu CracConfig was introduced very recently as a placeholder for SnapStart - before we can use that name in public. 

It does not remove the CRaC support, it just makes things less confusing, as if you would have used the SnapStart feature with another CRaC impl.... bad things would have happened.

**Pull request merged**